### PR TITLE
exec: handle some mixed type comparison expressions

### PR DESCRIPTION
**Pull request by rafiss**
This adds support for int-int comparisons and float-float comparisons
(of different sizes), as well as various string comparisons. This
is necessary to support TPC-H queries 7, 16, 19, and 21 (though there
are still other issues to address before those are fully supported).

There is some refactoring here so that as a next step we can support
comparisons between different types entirely (e.g. int-decimal).

touches #39189

Release note: None

### CODE REVIEW

**Review by rohany — COMMENTED** [review]
> Reviewed 7 of 16 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39464)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @rafiss, and @rohany)

___
*[pkg/sql/distsqlrun/column_exec_setup.go, line 911 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39464#-LlmSLqwEa4dMSsmzeCk:-LlmSLqwEa4dMSsmzeCl:bxg83tm) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7c418c30a549a14cd7abe84405447025f850fcc/pkg/sql/distsqlrun/column_exec_setup.go#L911)):*
> ```Go
> 		// handles invalid types for the IN and NOT IN operations at this point,
> 		// and we allow a comparison between t and t tuple.
> 		if t.Operator == tree.In || t.Operator == tree.NotIn {
> ```

I don't think we should remove this? I don't think the same rules apply for IN/NOT IN
___
*[pkg/sql/exec/execgen/cmd/execgen/overloads.go, line 373 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39464#-Llmj9Gf1l1WbUH4i_P1:-Llmj9Gf1l1WbUH4i_P2:bt5pewp) ([raw file](https://github.com/cockroachdb/cockroach/blob/446730db58ac56ee60ca802cc1c4a12b6a22c1a3/pkg/sql/exec/execgen/cmd/execgen/overloads.go#L373)):*
> ```Go
> }
> 
> func (c intCustomizer) getCmpOpCompareFunc() compareFunc {
> ```

is the idea that we add more functions like this + compatible types below to add support for mixed type operations?
___
*[pkg/sql/exec/compare.go, line 16 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39464#-LlmRrr4D-_bKoh5asPI:-LlmRrr4D-_bKoh5asPJ:bvvuqza) ([raw file](https://github.com/cockroachdb/cockroach/blob/446730db58ac56ee60ca802cc1c4a12b6a22c1a3/pkg/sql/exec/compare.go#L16)):*
> ```Go
> 
> // compareInts compares two int values. This function allows us to easily
> // handle mixec-type integer comparison.
> ```

[nit] misspelled

**Review by yuzefovich — COMMENTED** [review]
Nice work! A few nits from me.

> Reviewed 9 of 16 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39464#-:-Lm6cscO9xwdqe8pcILw:beecu6p)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @rafiss)

___
*[pkg/sql/distsqlrun/column_exec_setup.go, line 849 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39464#-Lm6_lmV0qgCDhlQXMIT:-Lm6_lmV0qgCDhlQXMIU:b-el22fq) ([raw file](https://github.com/cockroachdb/cockroach/blob/446730db58ac56ee60ca802cc1c4a12b6a22c1a3/pkg/sql/distsqlrun/column_exec_setup.go#L849)):*
> ```Go
> 	}
> 	if rConstArg, rConst := right.(tree.Datum); rConst {
> ```

[nit]: extra newline.
___
*[pkg/sql/exec/execgen/cmd/execgen/projection_ops_gen.go, line 221 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39464#-Lm6bdZwDQ3-L8RK-uhk:-Lm6bdZwDQ3-L8RK-uhl:b-5mog8) ([raw file](https://github.com/cockroachdb/cockroach/blob/446730db58ac56ee60ca802cc1c4a12b6a22c1a3/pkg/sql/exec/execgen/cmd/execgen/projection_ops_gen.go#L221)):*
> ```Go
> {{range $left := .ConstSides}}
> // GetProjectionConstOperator returns the appropriate constant projection
> // operator for the given column type and comparison.
> ```

[nit]: this comment needs minor update as well.
___
*[pkg/sql/exec/execgen/cmd/execgen/projection_ops_gen.go, line 289 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39464#-Lm6bi0a53wQUWjXCejx:-Lm6bi0a53wQUWjXCejy:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/446730db58ac56ee60ca802cc1c4a12b6a22c1a3/pkg/sql/exec/execgen/cmd/execgen/projection_ops_gen.go#L289)):*
> ```Go
> 
> // GetProjectionOperator returns the appropriate projection operator for the
> // given column type and comparison.
> ```

ditto
___
*[pkg/sql/exec/execgen/cmd/execgen/selection_ops_gen.go, line 253 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39464#-Lm6bzdP32_egMoQeLdX:-Lm6bzdP32_egMoQeLdY:b8ssyx1) ([raw file](https://github.com/cockroachdb/cockroach/blob/446730db58ac56ee60ca802cc1c4a12b6a22c1a3/pkg/sql/exec/execgen/cmd/execgen/selection_ops_gen.go#L253)):*
> ```Go
> 	typToOverloads := make(map[types.T][]*overload)
> 	for _, overload := range comparisonOpOverloads {
> 		if overload.LTyp != overload.RTyp {
> ```

[nit]: a comment why we're skipping mixed types here might be helpful.
___
*[pkg/sql/exec/types/types.go, line 66 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39464#-Lm6cIBx7Xa465AAzLZF:-Lm6cIBx7Xa465AAzLZG:bnrp086) ([raw file](https://github.com/cockroachdb/cockroach/blob/446730db58ac56ee60ca802cc1c4a12b6a22c1a3/pkg/sql/exec/types/types.go#L66)):*
> ```Go
> 	CompatibleTypes[Bytes] = append(CompatibleTypes[Bytes], Bytes)
> 	CompatibleTypes[Decimal] = append(CompatibleTypes[Decimal], Decimal)
> 	CompatibleTypes[Int8] = append(CompatibleTypes[Int8], Int8)
> ```

[nit]: it might be cleaner to define "all int types" and use that to append to compatible types.
___
*[pkg/sql/logictest/testdata/logic_test/vectorize, line 70 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39464#-Lm6cXkg0NVpUNlJKG5Y:-Lm6cXkg0NVpUNlJKG5Z:bkk5r4f) ([raw file](https://github.com/cockroachdb/cockroach/blob/446730db58ac56ee60ca802cc1c4a12b6a22c1a3/pkg/sql/logictest/testdata/logic_test/vectorize#L70)):*
> ```
> # Mixed type comparison
> query IB
> SELECT c, c > 1 FROM a LIMIT 3
> ```

[nit]: I think you might be able to cast an INT8 to INT4 here to make it explicit mixed-type comparison (then you'd also not need a new column). Additionally, it'd be nice to see comparison with other int types.

**Review by rafiss — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39464)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @rafiss, @rohany, and @yuzefovich)

___
*[pkg/sql/distsqlrun/column_exec_setup.go, line 849 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39464#-Lm6_lmV0qgCDhlQXMIT:-LmFlcviE-pXyy56waZb:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/446730db58ac56ee60ca802cc1c4a12b6a22c1a3/pkg/sql/distsqlrun/column_exec_setup.go#L849)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[nit]: extra newline.
</blockquote></details>

done
___
*[pkg/sql/distsqlrun/column_exec_setup.go, line 911 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39464#-LlmSLqwEa4dMSsmzeCk:-LmFlP0r4S05WGxqeqUU:b-4vti1t) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7c418c30a549a14cd7abe84405447025f850fcc/pkg/sql/distsqlrun/column_exec_setup.go#L911)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

I don't think we should remove this? I don't think the same rules apply for IN/NOT IN
</blockquote></details>

Not sure I follow -- previously we already were returning a `nil` error here in the case of IN/NOT IN. We will continue to do that.
___
*[pkg/sql/exec/execgen/cmd/execgen/overloads.go, line 373 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39464#-Llmj9Gf1l1WbUH4i_P1:-LmG726T9HOcBmYWvSBV:b-p2f6qz) ([raw file](https://github.com/cockroachdb/cockroach/blob/446730db58ac56ee60ca802cc1c4a12b6a22c1a3/pkg/sql/exec/execgen/cmd/execgen/overloads.go#L373)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

is the idea that we add more functions like this + compatible types below to add support for mixed type operations?
</blockquote></details>

yeah, that was my thinking. though to really support mixed types, we'd need to have `customizer`s that are like intDecimal (etc), not just between different ints.
___
*[pkg/sql/exec/execgen/cmd/execgen/projection_ops_gen.go, line 221 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39464#-Lm6bdZwDQ3-L8RK-uhk:-LmG7flFEKuUQULM4sj-:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/446730db58ac56ee60ca802cc1c4a12b6a22c1a3/pkg/sql/exec/execgen/cmd/execgen/projection_ops_gen.go#L221)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[nit]: this comment needs minor update as well.
</blockquote></details>

done
___
*[pkg/sql/exec/execgen/cmd/execgen/projection_ops_gen.go, line 289 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39464#-Lm6bi0a53wQUWjXCejx:-LmG7fAe2X2DCAOUipal:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/446730db58ac56ee60ca802cc1c4a12b6a22c1a3/pkg/sql/exec/execgen/cmd/execgen/projection_ops_gen.go#L289)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

ditto
</blockquote></details>

done
___
*[pkg/sql/exec/execgen/cmd/execgen/selection_ops_gen.go, line 253 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39464#-Lm6bzdP32_egMoQeLdX:-LmG8zbd8onalCFsZOZ6:bg0nnlv) ([raw file](https://github.com/cockroachdb/cockroach/blob/446730db58ac56ee60ca802cc1c4a12b6a22c1a3/pkg/sql/exec/execgen/cmd/execgen/selection_ops_gen.go#L253)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[nit]: a comment why we're skipping mixed types here might be helpful.
</blockquote></details>

actually, on further consideration, we should support mixed types here too
___
*[pkg/sql/exec/compare.go, line 16 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39464#-LlmRrr4D-_bKoh5asPI:-LmG6tnd112mdVw197J-:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/446730db58ac56ee60ca802cc1c4a12b6a22c1a3/pkg/sql/exec/compare.go#L16)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

[nit] misspelled
</blockquote></details>

done

**Review by rohany — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39464)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @rafiss, @rohany, and @yuzefovich)

___
*[pkg/sql/distsqlrun/column_exec_setup.go, line 911 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39464#-LlmSLqwEa4dMSsmzeCk:-LmGGIXQAFB7rj0u74mQ:b-ynn712) ([raw file](https://github.com/cockroachdb/cockroach/blob/e7c418c30a549a14cd7abe84405447025f850fcc/pkg/sql/distsqlrun/column_exec_setup.go#L911)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

Not sure I follow -- previously we already were returning a `nil` error here in the case of IN/NOT IN. We will continue to do that.
</blockquote></details>

Yeah never mind, I think I misunderstood.

**Review by yuzefovich — COMMENTED** [review]
> Reviewed 14 of 14 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39464)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @rafiss, @rohany, and @yuzefovich)

___
*[pkg/col/coltypes/types.go, line 66 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39464#-LmGUDVWBON9TIhmFC1G:-LmGUDVWBON9TIhmFC1H:bnrp086) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c6395ca34060630d69808066f8c4a6e8eae5c84/pkg/col/coltypes/types.go#L66)):*
> ```Go
> 	CompatibleTypes[Bytes] = append(CompatibleTypes[Bytes], Bytes)
> 	CompatibleTypes[Decimal] = append(CompatibleTypes[Decimal], Decimal)
> 	CompatibleTypes[Int8] = append(CompatibleTypes[Int8], Int8)
> ```

[nit]: it might be cleaner to define "all int types" and use that to append to compatible types.
___
*[pkg/sql/distsqlrun/column_exec_setup.go, line 915 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39464#-LmGTWwHDNttyHcnxM6q:-LmGTWwHDNttyHcnxM6r:bb8x21w) ([raw file](https://github.com/cockroachdb/cockroach/blob/42190f70f8fdcd67eadd64f3a189aa3f0e6cb9cd/pkg/sql/distsqlrun/column_exec_setup.go#L915)):*
> ```Go
> 		}
> 
> 		if !left.Identical(right) {
> ```

I think this part needs to be kept though, right?

**Review by rafiss — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39464)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @rafiss, @rohany, and @yuzefovich)

___
*[pkg/sql/distsqlrun/column_exec_setup.go, line 915 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39464#-LmGTWwHDNttyHcnxM6q:-LmGWYvj8Un3SFG72fFR:bon3ky2) ([raw file](https://github.com/cockroachdb/cockroach/blob/42190f70f8fdcd67eadd64f3a189aa3f0e6cb9cd/pkg/sql/distsqlrun/column_exec_setup.go#L915)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I think this part needs to be kept though, right?
</blockquote></details>

This part is removed because we will now allow comparisons to happen even if the left and right type are not identical.

**Review by rafiss — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39464)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @rafiss, @rohany, and @yuzefovich)

___
*[pkg/col/coltypes/types.go, line 66 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39464#-LmGUDVWBON9TIhmFC1G:-LmG_xDsAo-yDgIChzdH:b-90vxgq) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c6395ca34060630d69808066f8c4a6e8eae5c84/pkg/col/coltypes/types.go#L66)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[nit]: it might be cleaner to define "all int types" and use that to append to compatible types.
</blockquote></details>

yeah i like that idea. i'll do it

**Comment by rafiss** [comment]
This is ready for another review, whenever someone has a chance

**Review by yuzefovich — APPROVED** [review]
:lgtm: but I haven't thought much about mixed-type comparisons, so ideally I'd wait for Rohan's approval.

> Reviewed 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39464#-:-LmKl3RPBAJ4qTS4wFOU:bmvhy42)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis, @rafiss, @rohany, and @yuzefovich)

___
*[pkg/sql/distsqlrun/column_exec_setup.go, line 915 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/39464#-LmGTWwHDNttyHcnxM6q:-LmKkqFl68EQFZmFHSrg:bkrelnw) ([raw file](https://github.com/cockroachdb/cockroach/blob/42190f70f8fdcd67eadd64f3a189aa3f0e6cb9cd/pkg/sql/distsqlrun/column_exec_setup.go#L915)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

This part is removed because we will now allow comparisons to happen even if the left and right type are not identical.
</blockquote></details>

Ok, SGTM.

**Review by rohany — APPROVED** [review]
This LGTM, should probably work with matt next to get tests that stress the mixed type operations here.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39464#-:-LmLGOKt1ptu25mamgSP:b-3gvdic)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis, @rafiss, @rohany, and @yuzefovich)

**Comment by rafiss** [comment]
thanks all!

bors r+

**Pull request merged**
